src/expr/src/scalar/func/impls/jsonb.rs

```diff
@@ -0,0 +1,180 @@
+            datum => Err(EvalError::InvalidJsonbCast {
+                from: jsonb_type(datum).into(),
+                to: "bigint".into(),
+            }),
+        }
+    }
+);
+
+sqlfunc!(
+    #[sqlname = "jsonbtof32"]
+    fn cast_jsonb_to_float32<'a>(a: JsonbRef<'a>) -> Result<f32, EvalError> {
+        match a.into_datum() {
+            Datum::Numeric(a) => cast_numeric_to_float32(a.into_inner()),
+            datum => Err(EvalError::InvalidJsonbCast {
+                from: jsonb_type(datum).into(),
+                to: "real".into(),
+            }),
+        }
+    }
+);
+
+sqlfunc!(
+    #[sqlname = "jsonbtof64"]
+    fn cast_jsonb_to_float64<'a>(a: JsonbRef<'a>) -> Result<f64, EvalError> {
+        match a.into_datum() {
+            Datum::Numeric(a) => cast_numeric_to_float64(a.into_inner()),
+            datum => Err(EvalError::InvalidJsonbCast {
+                from: jsonb_type(datum).into(),
+                to: "double precision".into(),
+            }),
+        }
+    }
+);
+
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
+pub struct CastJsonbToNumeric(pub Option<NumericMaxScale>);
+
+impl<'a> EagerUnaryFunc<'a> for CastJsonbToNumeric {
+    type Input = JsonbRef<'a>;
+    type Output = Result<Numeric, EvalError>;
+
+    fn call(&self, a: JsonbRef<'a>) -> Result<Numeric, EvalError> {
+        match a.into_datum() {
+            Datum::Numeric(mut num) => match self.0 {
+                None => Ok(num.into_inner()),
+                Some(scale) => {
+                    if numeric::rescale(&mut num.0, scale.into_u8()).is_err() {
+                        return Err(EvalError::NumericFieldOverflow);
+                    };
+                    Ok(num.into_inner())
+                }
+            },
+            datum => Err(EvalError::InvalidJsonbCast {
+                from: jsonb_type(datum).into(),
+                to: "numeric".into(),
+            }),
+        }
+    }
+
+    fn output_type(&self, input: ColumnType) -> ColumnType {
+        ScalarType::Numeric { max_scale: self.0 }.nullable(input.nullable)
+    }
+}
+
+impl fmt::Display for CastJsonbToNumeric {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("jsonbtonumeric")
+    }
+}
+
+sqlfunc!(
+    #[sqlname = "jsonbtobool"]
+    fn cast_jsonb_to_bool<'a>(a: JsonbRef<'a>) -> Result<bool, EvalError> {
+        match a.into_datum() {
+            Datum::True => Ok(true),
+            Datum::False => Ok(false),
+            datum => Err(EvalError::InvalidJsonbCast {
+                from: jsonb_type(datum).into(),
+                to: "boolean".into(),
+            }),
+        }
+    }
+);
+
+sqlfunc!(
+    #[sqlname = "jsonb?tojsonb"]
+    fn cast_jsonb_or_null_to_jsonb<'a>(a: Option<JsonbRef<'a>>) -> JsonbRef<'a> {
+        match a.map(|v| v.into_datum()) {
+            None => JsonbRef::from_datum(Datum::JsonNull),
+            Some(Datum::Numeric(n)) => {
```

Comment:
@sploiselle this function is supposed to convert jsonb or null to jsonb but it handles more than the null case. Why do we do all these checks for finiteness and nans?